Multi version and `no-std`

la10736 opened the pull request:
Support multiple risc0 version in `no-std`

Most of the code come from `risc0-zkvm` crate: just removed useless implementations.

In `generate_proofs` there is a program to generate method and proofs. The `header-fix` folder conatins a simple script to mix license headers from risc0 and our..

Add a test for no-std.

la10736 committed:
Support multiple risc0 version in `no-std`

la10736 committed:
Use cargo make also for udeps (make ci coherent).

Add cases for 1.2.0-1.1.3 matrix cell.

Simplified how to select risc0 version (toml and docs)

la10736 committed:
Recover old broken test methods

95DDB commented on src/context.rs:
```diff
@@ -0,0 +1,124 @@
+    segment::SegmentReceiptVerifierParameters,
+};
+
+/// Context available to the verification process.
```
Maybe add here or in the readme some rationale about the choice of the kind of verifiers we support

95DDB commented on src/receipt.rs:
```diff
@@ -0,0 +1,283 @@
+
+    /// Verify the prove against the given public inputs `pubs` and the vk `image_id`.  
+    ///
+    /// The context version used is the last `v1_1` that is compatible with all proof generated
+    /// with Risc0 1.1.x versions. If you are interested to verify older version see
+    /// `verify_with_context` instead.
```
Outdated comment

95DDB commented on src/receipt.rs:
```diff
@@ -0,0 +1,283 @@
+    /// [ReceiptClaim]. This method additionally ensures that the guest exited with a successful
+    /// status code (i.e. `Halted(0)`), the image ID is as expected, and the journal has not been
+    /// tampered with.
+    pub fn verify_with_context<SC: CircuitCoreDef, RC: CircuitCoreDef>(
```
Probably worth, to avoid issues, to keep only a verify function and asking to explicitly pass the verifier context each time

la10736 replied:
I was thinking exactly the same thing... these two function just add confusion and is not clear what context you sue in `verify()`: I'm just renaming `verify_with_context()` into `verify`

95DDB reviewed (commented):
LGTM amazing job ! A couple of requests:

- Document how to update the code in case of new Risc0 upgrades
- Add a small example on how to parse and verify Risc0 artifacts in the main README.md
- Fix typos in readmes and docs

la10736 committed:
- Add tests to check if it's possible for an external user replace the hash implementation.
- Add last matrix test cases
- little docs fix up

lgiussan commented on src/receipt/succinct.rs:
```diff
@@ -0,0 +1,319 @@
+            digest!("f171d19df8f27878677080c5e4c38ed2655f5f54302468ce805594a4b3e38104")
+        );
+    }
+}
```
```suggestion
#[cfg(test)]
mod tests {

    use super::SuccinctReceiptVerifierParameters;
    use crate::sha::Digestible;
    use risc0_zkp::core::digest::{digest, Digest};
    use rstest::rstest;

    // Check that the verifier parameters has a stable digest (and therefore a stable value). This
    // struct encodes parameters used in verification, and so this value should be updated if and
    // only if a change to the verifier parameters is expected. Updating the verifier parameters
    // will result in incompatibility with previous versions.
    #[rstest]
    #[case::v1_0(SuccinctReceiptVerifierParameters::v1_0().digest(), digest!("f171d19df8f27878677080c5e4c38ed2655f5f54302468ce805594a4b3e38104"))]
    #[case::v1_1(SuccinctReceiptVerifierParameters::v1_1().digest(), digest!("71023badfee05b76de871c5cc5a95cbedf50395e3634ffb9f3192950b16a77ae"))]
    #[case::v1_2(SuccinctReceiptVerifierParameters::v1_2().digest(), digest!("21a829e931cda9f34723dc77d947efe264771fea83bc495b3903014d0fe50d57"))]
    fn succinct_receipt_verifier_parameters_is_stable(
        #[case] computed: Digest,
        #[case] hardcoded: Digest,
    ) {
        assert_eq!(computed, hardcoded);
    }
}
```

lgiussan commented on src/circuit.rs:
```diff
@@ -0,0 +1,245 @@
+}
+
+pub mod v1_2 {
+    pub use risc0_circuit_rv32im::*;
```
For the sake of uniformity, we could treat also the latest version in the same manner as the previous ones (i.e. copy-pasting the circuits inside of `src/circuit/v1_2`, instead of importing the `risc0_circuit_rv32im` and `risc0_circuit_recursion` crates)?

lgiussan reviewed (approved):
LGTM!

lgiussan commented on src/deserializer.rs:
Please also remove the (now unused) `src/proof.rs` file

la10736 reviewed (commented):
> For the sake of uniformity, we could treat also the latest version in the same manner as the previous ones (i.e. copy-pasting the circuits inside of `src/circuit/v1_2`, instead of importing the `risc0_circuit_rv32im` and `risc0_circuit_recursion` crates)? 

I was fought about what's the best approach  here. Anyway maybe the actula one is little bit fragile because we need to use `=` in cargo toml to avoid unexpected updataing. So, I'll accept your subbestion

la10736 committed:
Make the code more coherent and included also the 1.2 version in the local source.
- Some refactor
- removed unused files

la10736 committed:
Removed verify_with_context and clean up some documentation

la10736 committed:
- Document how to introduce the new version
- Add the method used to create the tests
- Add test for segment context stability
